File: maximum-difference-between-node-and-ancestor/maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        stack = [(root, float('-inf'), float('inf'), float('-inf'))]
        ans = 0
        while stack:
            node, mx, mn, max_diff = stack.pop()
            mx, mn = max(mx, node.val), min(mn, node.val)
            max_diff = max(abs(mx - mn), max_diff)
            ans = max(ans, max_diff)
            
            if node.left:
                stack.append((node.left, mx, mn, max_diff))
            
            if node.right:
                stack.append((node.right, mx, mn, max_diff))
            
        return ans

File: maximum-difference-between-node-and-ancestor/maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxAncestorDiff(TreeNode* root) {
        int ans = 0;
        stack<pair<TreeNode*, pair<int ,int>>> *st = new stack<pair<TreeNode*, pair<int, int>>>;
        st -> push({root, {INT_MIN, INT_MAX}});
        while (!st -> empty()){
            auto it = st -> top();
            TreeNode* node = it.first;
            int mx = it.second.first, mn = it.second.second;
            st -> pop();
            mx = max(mx, node -> val);
            mn = min(mn, node -> val);
            ans = max(abs(mx - mn), ans);
            
            if (node -> left) st -> push({node -> left, {mx, mn}});
            if (node -> right) st -> push({node -> right, {mx, mn}});
        }
        
        return ans;
    }
};